PR failed IT on hammerbox:

https://gitlab.prod.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/20974

```
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest.test_indexing)
----------------------------------------------------------------------
Traceback (most recent call last):
…
  File "/builds/ucsc/azul/src/azul/queues.py", line 270, in wait_to_stabilize
    raise Exception('The queues have stalled', final_length)
Exception: ('The queues have stalled', 7)
----------------------------------------------------------------------
```

Hammebox fail queue: 
[azul-notifications_fail-hammerbox.json](https://github.com/DataBiosphere/azul/files/14101653/azul-notifications_fail-hammerbox.json)


[CloudWatch logs](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:logs-insights$3FqueryDetail$3D~(end~'2024-01-30T09*3a39*3a00.000Z~start~'2024-01-30T08*3a53*3a41.428Z~timeType~'ABSOLUTE~tz~'UTC~editorString~'fields*20*40timestamp*2c*20*40message*0a*7c*20filter*20*40message*20like*20*2fERROR*2f*0a*7c*20sort*20*40timestamp*20desc*0a*7c*20limit*20200~queryId~'c5bcdb3580a85712-403f2238-4f76bcd-9e8daa5-8259a5388a9222a2a7a7a~source~(~'*2faws*2flambda*2fazul-indexer-hammerbox-aggregate~'*2faws*2flambda*2fazul-indexer-hammerbox-contribute~'*2faws*2flambda*2fazul-indexer-hammerbox-contribute_retry~'*2faws*2flambda*2fazul-indexer-hammerbox-indexercachehealth~'*2faws*2flambda*2fazul-indexer-hammerbox))):

```
[ERROR] RequirementError
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 166, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1756, in __call__
    return self.handler(event_obj)
  File "/var/task/app.py", line 337, in contribute_retry
    app.index_controller.contribute(event, retry=True)
  File "/var/task/azul/indexer/index_controller.py", line 180, in contribute
    contributions, replicas = self.transform(catalog, notification, delete)
  File "/var/task/azul/indexer/index_controller.py", line 228, in transform
    bundle = service.fetch_bundle(catalog, bundle_fqid)
  File "/var/task/azul/indexer/index_service.py", line 193, in fetch_bundle
    return plugin.fetch_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr.py", line 212, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 288, in _emulate_bundle
    return self._primary_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr_anvil/__init__.py", line 339, in _primary_bundle
    require(row.pop('part_of_dataset_id') == dataset_id)
  File "/var/task/azul/__init__.py", line 1531, in require
    reject(not condition, *args, exception=exception)
  File "/var/task/azul/__init__.py", line 1550, in reject
    raise exception(*args)
```

--- a/src/azul/plugins/repository/tdr_anvil/__init__.py
+++ b/src/azul/plugins/repository/tdr_anvil/__init__.py
@@ -328,14 +328,16 @@ class Plugin(TDRPlugin[TDRAnvilBundle, TDRSourceSpec, TDRSourceRef, TDRAnvilBund
         for entity_type, typed_keys in sorted(keys_by_type.items()):
             pk_column = entity_type + '_id'
             rows = self._retrieve_entities(source.spec, entity_type, typed_keys)
-            if entity_type == 'donors':
+            if entity_type == 'donor':
                 # We expect that the foreign key `part_of_dataset_id` is
                 # redundant for biosamples and donors. To simplify our queries,
                 # we do not follow the latter during the graph traversal.
-                # Here, we validate our expectation.
-                dataset_id: Key = one(keys_by_type['datasets'])
+                # Here, we validate our expectation. Note that the key is an
+                # array for biosamples, but not for donors.
+                dataset_id: Key = one(keys_by_type['dataset'])
                 for row in rows:
-                    require(row.pop('part_of_dataset_id') == dataset_id)
+                    donor_dataset_id = row.pop('part_of_dataset_id')
+                    require(donor_dataset_id == dataset_id, donor_dataset_id, dataset_id)
             for row in sorted(rows, key=itemgetter(pk_column)):
                 key = KeyReference(key=row[pk_column], entity_type=entity_type)
                 entity = EntityReference(entity_id=row['datarepo_row_id'], entity_type=entity_type)

--- a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json
+++ b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.results.json
@@ -863,7 +863,7 @@
     {
         "_index": "azul_v2_nadove4_test_replica",
         "_type": "_doc",
-        "_id": "15b76f9c-6b46-433f-851d-34e89f1b9ba6_ab03c58f05fb6bf72860fee14821f1feb9e331cc",
+        "_id": "15b76f9c-6b46-433f-851d-34e89f1b9ba6_ed0a841f18d24d3955ace40a353ad640e1da60f1",
         "_score": 1.0,
         "_source": {
             "entity_id": "15b76f9c-6b46-433f-851d-34e89f1b9ba6",
@@ -876,7 +876,7 @@
                 "file_id": "1e269f04-4347-4188-b060-1dcc69e71d67",
                 "file_md5sum": "vuxgbuCqKZ/fkT9CWTFmIg==",
                 "file_name": "307500.merged.matefixed.sorted.markeddups.recal.g.vcf.gz",
-                "file_ref": "drs://data.terra.bio/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_1e269f04-4347-4188-b060-1dcc69e71d67",
+                "file_ref": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_1e269f04-4347-4188-b060-1dcc69e71d67",
                 "file_size": 213021639,
                 "is_supplementary": false,
                 "reference_assembly": [],
@@ -893,7 +893,7 @@
     {
         "_index": "azul_v2_nadove4_test_replica",
         "_type": "_doc",
-        "_id": "15d85d30-ad4a-4f50-87a8-a27f59dd1b5f_bb4d58a1aa21adce6af1b09e7e75664afda5890e",
+        "_id": "15d85d30-ad4a-4f50-87a8-a27f59dd1b5f_92347daeeff2cbb94cd3bf9f73bb40e8c52a1e40",
         "_score": 1.0,
         "_source": {
             "entity_id": "15d85d30-ad4a-4f50-87a8-a27f59dd1b5f",
@@ -918,7 +918,7 @@
                 "source_datarepo_row_ids": [
                     "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
                 ],
-                "version": "redacted-/JIktcFZ"
+                "version": "2022-06-01T00:00:00.000000Z"
             }
         }
     },
@@ -1721,6 +1721,36 @@
         }
     },
     {
+        "_index": "azul_v2_nadove4_test_replica",
+        "_type": "_doc",
+        "_id": "3b17377b-16b1-431c-9967-e5d01fc5923f_33d8fab596f5af223936ba8f7ef6d218e4932cde",
+        "_score": 1.0,
+        "_source": {
+            "entity_id": "3b17377b-16b1-431c-9967-e5d01fc5923f",
+            "contents": {
+                "data_modality": [],
+                "datarepo_row_id": "3b17377b-16b1-431c-9967-e5d01fc5923f",
+                "file_format": ".bam",
+                "file_id": "8b722e88-8103-49c1-b351-e64fa7c6ab37",
+                "file_md5sum": "fNn9e1SovzgOROk3BvH6LQ==",
+                "file_name": "307500.merged.matefixed.sorted.markeddups.recal.bam",
+                "file_ref": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
+                "file_size": 3306845592,
+                "is_supplementary": false,
+                "reference_assembly": [],
+                "source_datarepo_row_ids": [
+                    "file_inventory:9658d94a-511d-4b49-82c3-d0cb07e0cff2"
+                ],
+                "version": "2022-06-01T00:00:00.000000Z",
+                "drs_uri": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
+                "sha256": "",
+                "crc32": ""
+            },
+            "replica_type": "anvil_file",
+            "hub_ids": []
+        }
+    },
+    {
         "_index": "azul_v2_nadove4_test_files",
         "_type": "_doc",
         "_id": "3b17377b-16b1-431c-9967-e5d01fc5923f_826dea02-e274-affe-aabc-eb3db63ad068__exists",
@@ -1906,36 +1936,6 @@
             "bundle_uuid": "826dea02-e274-affe-aabc-eb3db63ad068",
             "bundle_version": "",
             "bundle_deleted": false
-        }
-    },
-    {
-        "_index": "azul_v2_nadove4_test_replica",
-        "_type": "_doc",
-        "_id": "3b17377b-16b1-431c-9967-e5d01fc5923f_da659bb0d46d83e5fd038f119f6fb76b0fc46b6e",
-        "_score": 1.0,
-        "_source": {
-            "entity_id": "3b17377b-16b1-431c-9967-e5d01fc5923f",
-            "replica_type": "anvil_file",
-            "hub_ids": [],
-            "contents": {
-                "data_modality": [],
-                "datarepo_row_id": "3b17377b-16b1-431c-9967-e5d01fc5923f",
-                "file_format": ".bam",
-                "file_id": "8b722e88-8103-49c1-b351-e64fa7c6ab37",
-                "file_md5sum": "fNn9e1SovzgOROk3BvH6LQ==",
-                "file_name": "307500.merged.matefixed.sorted.markeddups.recal.bam",
-                "file_ref": "drs://data.terra.bio/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
-                "file_size": 3306845592,
-                "is_supplementary": false,
-                "reference_assembly": [],
-                "source_datarepo_row_ids": [
-                    "file_inventory:9658d94a-511d-4b49-82c3-d0cb07e0cff2"
-                ],
-                "version": "2022-06-01T00:00:00.000000Z",
-                "drs_uri": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
-                "sha256": "",
-                "crc32": ""
-            }
         }
     },
     {
@@ -2930,7 +2930,7 @@
     {
         "_index": "azul_v2_nadove4_test_replica",
         "_type": "_doc",
-        "_id": "826dea02-e274-4ffe-aabc-eb3db63ad068_df5c9e8f2fdbbd0f85316396c1119451830f332f",
+        "_id": "826dea02-e274-4ffe-aabc-eb3db63ad068_a990d631b61245811f3401dfa59a2d47edac01a5",
         "_score": 1.0,
         "_source": {
             "entity_id": "826dea02-e274-4ffe-aabc-eb3db63ad068",
@@ -2949,7 +2949,7 @@
                 "source_datarepo_row_ids": [
                     "sample:98048c3b-2525-4090-94fd-477de31f2608"
                 ],
-                "version": "redacted-/JIktcFZ"
+                "version": "2022-06-01T00:00:00.000000Z"
             }
         }
     },
@@ -3506,7 +3506,7 @@
     {
         "_index": "azul_v2_nadove4_test_replica",
         "_type": "_doc",
-        "_id": "939a4bd3-86ed-4a8a-81f4-fbe0ee673461_026c212506d50c06d8225c2d5fac1e789eb4d301",
+        "_id": "939a4bd3-86ed-4a8a-81f4-fbe0ee673461_0a85e766a03e5aedfa1bbcf0932e376dba5cd431",
         "_score": 1.0,
         "_source": {
             "entity_id": "939a4bd3-86ed-4a8a-81f4-fbe0ee673461",
@@ -3531,7 +3531,7 @@
                 "source_datarepo_row_ids": [
                     "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
                 ],
-                "version": "redacted-/JIktcFZ"
+                "version": "2022-06-01T00:00:00.000000Z"
             }
         }
     },
@@ -3844,31 +3844,6 @@
         }
     },
     {
-        "_index": "azul_v2_nadove4_test_replica",
-        "_type": "_doc",
-        "_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e_1d8e455a019e942e4d2298f5f019778164640246",
-        "_score": 1.0,
-        "_source": {
-            "entity_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e",
-            "replica_type": "anvil_donor",
-            "hub_ids": [],
-            "contents": {
-                "datarepo_row_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e",
-                "donor_id": "1e2bd7e5-f45e-a391-daea-7c060be76acd",
-                "genetic_ancestry": [],
-                "organism_type": "redacted-ACw+6ecI",
-                "phenotypic_sex": "redacted-JfQ0b3xG",
-                "reported_ethnicity": [
-                    "redacted-NSkwDycK"
-                ],
-                "source_datarepo_row_ids": [
-                    "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
-                ],
-                "version": "redacted-/JIktcFZ"
-            }
-        }
-    },
-    {
         "_index": "azul_v2_nadove4_test_donors",
         "_type": "_doc",
         "_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e_826dea02-e274-affe-aabc-eb3db63ad068__exists",
@@ -4100,6 +4075,31 @@
             "bundle_uuid": "826dea02-e274-affe-aabc-eb3db63ad068",
             "bundle_version": "",
             "bundle_deleted": false
+        }
+    },
+    {
+        "_index": "azul_v2_nadove4_test_replica",
+        "_type": "_doc",
+        "_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e_ef7dbf88444d1f5f24e887821dbae7d1dc182e55",
+        "_score": 1.0,
+        "_source": {
+            "entity_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e",
+            "contents": {
+                "datarepo_row_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e",
+                "donor_id": "1e2bd7e5-f45e-a391-daea-7c060be76acd",
+                "genetic_ancestry": [],
+                "organism_type": "redacted-ACw+6ecI",
+                "phenotypic_sex": "redacted-JfQ0b3xG",
+                "reported_ethnicity": [
+                    "redacted-NSkwDycK"
+                ],
+                "source_datarepo_row_ids": [
+                    "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
+                ],
+                "version": "2022-06-01T00:00:00.000000Z"
+            },
+            "replica_type": "anvil_donor",
+            "hub_ids": []
         }
     }
 ]

--- a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tables.tdr.json
+++ b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tables.tdr.json
@@ -1,0 +1,240 @@
+{
+    "tables": {
+        "activity": {
+            "rows": [
+                {
+                    "activity_id": "",
+                    "activity_type": "",
+                    "generated_file_id": [],
+                    "source_datarepo_row_ids": [],
+                    "used_biosample_id": [],
+                    "used_file_id": []
+                }
+            ]
+        },
+        "alignmentactivity": {
+            "rows": [
+                {
+                    "activity_type": "",
+                    "alignmentactivity_id": "",
+                    "data_modality": [],
+                    "generated_file_id": [],
+                    "reference_assembly": [],
+                    "source_datarepo_row_ids": [],
+                    "used_file_id": []
+                }
+            ]
+        },
+        "assayactivity": {
+            "rows": [
+                {
+                    "activity_type": "",
+                    "antibody_id": [],
+                    "assay_type": "",
+                    "assayactivity_id": "",
+                    "data_modality": [],
+                    "generated_file_id": [],
+                    "source_datarepo_row_ids": [],
+                    "used_biosample_id": []
+                }
+            ]
+        },
+        "biosample": {
+            "rows": [
+                {
+                    "anatomical_site": null,
+                    "apriori_cell_type": [],
+                    "biosample_id": "f9d40cf6-37b8-22f3-ce35-0dc614d2452b",
+                    "biosample_type": null,
+                    "datarepo_row_id": "826dea02-e274-4ffe-aabc-eb3db63ad068",
+                    "disease": null,
+                    "donor_age_at_collection_lower_bound": null,
+                    "donor_age_at_collection_unit": null,
+                    "donor_age_at_collection_upper_bound": null,
+                    "donor_id": [
+                        "1e2bd7e5-f45e-a391-daea-7c060be76acd"
+                    ],
+                    "part_of_dataset_id": [
+                        "52ee7665-7033-63f2-a8d9-ce8e32666739"
+                    ],
+                    "source_datarepo_row_ids": [
+                        "sample:98048c3b-2525-4090-94fd-477de31f2608"
+                    ]
+                }
+            ]
+        },
+        "dataset": {
+            "rows": [
+                {
+                    "consent_group": [
+                        "DS-BDIS"
+                    ],
+                    "data_modality": [],
+                    "data_use_permission": [
+                        "DS-BDIS"
+                    ],
+                    "datarepo_row_id": "2370f948-2783-4eb6-afea-e022897f4dcf",
+                    "dataset_id": "52ee7665-7033-63f2-a8d9-ce8e32666739",
+                    "owner": [
+                        "Debbie Nickerson"
+                    ],
+                    "principal_investigator": [],
+                    "registered_identifier": [
+                        "phs000693"
+                    ],
+                    "source_datarepo_row_ids": [
+                        "workspace_attributes:7a22b629-9d81-4e4d-9297-f9e44ed760bc"
+                    ],
+                    "title": "ANVIL_CMG_UWASH_DS_BDIS"
+                }
+            ]
+        },
+        "diagnosis": {
+            "rows": [
+                {
+                    "datarepo_row_id": "15d85d30-ad4a-4f50-87a8-a27f59dd1b5f",
+                    "diagnosis_age_lower_bound": null,
+                    "diagnosis_age_unit": null,
+                    "diagnosis_age_upper_bound": null,
+                    "diagnosis_id": "5ebe9bc4-a1be-0ddf-7277-b1e88276d0f6",
+                    "disease": [
+                        "redacted-A61iJlLx"
+                    ],
+                    "donor_id": "1e2bd7e5-f45e-a391-daea-7c060be76acd",
+                    "onset_age_lower_bound": null,
+                    "onset_age_unit": null,
+                    "onset_age_upper_bound": null,
+                    "phenopacket": [],
+                    "phenotype": [
+                        "redacted-acSYHZUr"
+                    ],
+                    "source_datarepo_row_ids": [
+                        "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
+                    ]
+                },
+                {
+                    "datarepo_row_id": "939a4bd3-86ed-4a8a-81f4-fbe0ee673461",
+                    "diagnosis_age_lower_bound": null,
+                    "diagnosis_age_unit": null,
+                    "diagnosis_age_upper_bound": null,
+                    "diagnosis_id": "25ff8d32-18c9-fc3e-020a-5de20d35d906",
+                    "disease": [
+                        "redacted-g50ublm/"
+                    ],
+                    "donor_id": "1e2bd7e5-f45e-a391-daea-7c060be76acd",
+                    "onset_age_lower_bound": null,
+                    "onset_age_unit": null,
+                    "onset_age_upper_bound": null,
+                    "phenopacket": [],
+                    "phenotype": [
+                        "redacted-acSYHZUr"
+                    ],
+                    "source_datarepo_row_ids": [
+                        "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
+                    ]
+                }
+            ]
+        },
+        "donor": {
+            "rows": [
+                {
+                    "datarepo_row_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e",
+                    "donor_id": "1e2bd7e5-f45e-a391-daea-7c060be76acd",
+                    "genetic_ancestry": [],
+                    "organism_type": "redacted-ACw+6ecI",
+                    "part_of_dataset_id": "52ee7665-7033-63f2-a8d9-ce8e32666739",
+                    "phenotypic_sex": "redacted-JfQ0b3xG",
+                    "reported_ethnicity": [
+                        "redacted-NSkwDycK"
+                    ],
+                    "source_datarepo_row_ids": [
+                        "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
+                    ]
+                }
+            ]
+        },
+        "file": {
+            "rows": [
+                {
+                    "data_modality": [],
+                    "datarepo_row_id": "15b76f9c-6b46-433f-851d-34e89f1b9ba6",
+                    "file_format": ".vcf.gz",
+                    "file_id": "1e269f04-4347-4188-b060-1dcc69e71d67",
+                    "file_md5sum": "vuxgbuCqKZ/fkT9CWTFmIg==",
+                    "file_name": "307500.merged.matefixed.sorted.markeddups.recal.g.vcf.gz",
+                    "file_ref": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_1e269f04-4347-4188-b060-1dcc69e71d67",
+                    "file_size": 213021639,
+                    "is_supplementary": false,
+                    "reference_assembly": [],
+                    "source_datarepo_row_ids": [
+                        "file_inventory:81d16471-97ac-48fe-99a0-73d9ec62c2c0"
+                    ]
+                },
+                {
+                    "data_modality": [],
+                    "datarepo_row_id": "3b17377b-16b1-431c-9967-e5d01fc5923f",
+                    "file_format": ".bam",
+                    "file_id": "8b722e88-8103-49c1-b351-e64fa7c6ab37",
+                    "file_md5sum": "fNn9e1SovzgOROk3BvH6LQ==",
+                    "file_name": "307500.merged.matefixed.sorted.markeddups.recal.bam",
+                    "file_ref": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
+                    "file_size": 3306845592,
+                    "is_supplementary": false,
+                    "reference_assembly": [],
+                    "source_datarepo_row_ids": [
+                        "file_inventory:9658d94a-511d-4b49-82c3-d0cb07e0cff2"
+                    ]
+                }
+            ]
+        },
+        "sequencingactivity": {
+            "rows": [
+                {
+                    "activity_type": "Sequencing",
+                    "assay_type": [],
+                    "data_modality": [],
+                    "datarepo_row_id": "1509ef40-d1ba-440d-b298-16b7c173dcd4",
+                    "generated_file_id": [
+                        "1e269f04-4347-4188-b060-1dcc69e71d67"
+                    ],
+                    "sequencingactivity_id": "18b3be87-e26b-4376-0d8d-c1e370e90e07",
+                    "source_datarepo_row_ids": [
+                        "sequencing:d4f6c0c4-1e11-438e-8218-cfea63b8b051"
+                    ],
+                    "used_biosample_id": [
+                        "f9d40cf6-37b8-22f3-ce35-0dc614d2452b"
+                    ]
+                },
+                {
+                    "activity_type": "Sequencing",
+                    "assay_type": [],
+                    "data_modality": [],
+                    "datarepo_row_id": "816e364e-1193-4e5b-a91a-14e4b009157c",
+                    "generated_file_id": [
+                        "8b722e88-8103-49c1-b351-e64fa7c6ab37"
+                    ],
+                    "sequencingactivity_id": "a60c5138-3749-f7cb-8714-52d389ad5231",
+                    "source_datarepo_row_ids": [
+                        "sequencing:a6c663c7-6f26-4ed2-af9d-48e9c709a22b"
+                    ],
+                    "used_biosample_id": [
+                        "f9d40cf6-37b8-22f3-ce35-0dc614d2452b"
+                    ]
+                }
+            ]
+        },
+        "variantcallingactivity": {
+            "rows": [
+                {
+                    "activity_type": "",
+                    "data_modality": [],
+                    "generated_file_id": [],
+                    "reference_assembly": [],
+                    "source_datarepo_row_ids": [],
+                    "used_file_id": [],
+                    "variantcallingactivity_id": ""
+                }
+            ]
+        }
+    }
+}

--- a/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json
+++ b/test/indexer/data/826dea02-e274-affe-aabc-eb3db63ad068.tdr.anvil.json
@@ -13,7 +13,7 @@
             "source_datarepo_row_ids": [
                 "sample:98048c3b-2525-4090-94fd-477de31f2608"
             ],
-            "version": "redacted-/JIktcFZ"
+            "version": "2022-06-01T00:00:00.000000Z"
         },
         "dataset/2370f948-2783-4eb6-afea-e022897f4dcf": {
             "consent_group": [
@@ -57,7 +57,7 @@
             "source_datarepo_row_ids": [
                 "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
             ],
-            "version": "redacted-/JIktcFZ"
+            "version": "2022-06-01T00:00:00.000000Z"
         },
         "diagnosis/939a4bd3-86ed-4a8a-81f4-fbe0ee673461": {
             "datarepo_row_id": "939a4bd3-86ed-4a8a-81f4-fbe0ee673461",
@@ -78,7 +78,7 @@
             "source_datarepo_row_ids": [
                 "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
             ],
-            "version": "redacted-/JIktcFZ"
+            "version": "2022-06-01T00:00:00.000000Z"
         },
         "donor/bfd991f2-2797-4083-972a-da7c6d7f1b2e": {
             "datarepo_row_id": "bfd991f2-2797-4083-972a-da7c6d7f1b2e",
@@ -92,7 +92,7 @@
             "source_datarepo_row_ids": [
                 "subject:c23887a0-20c1-44e4-a09e-1c5dfdc2d0ef"
             ],
-            "version": "redacted-/JIktcFZ"
+            "version": "2022-06-01T00:00:00.000000Z"
         },
         "file/15b76f9c-6b46-433f-851d-34e89f1b9ba6": {
             "data_modality": [],
@@ -101,7 +101,7 @@
             "file_id": "1e269f04-4347-4188-b060-1dcc69e71d67",
             "file_md5sum": "vuxgbuCqKZ/fkT9CWTFmIg==",
             "file_name": "307500.merged.matefixed.sorted.markeddups.recal.g.vcf.gz",
-            "file_ref": "drs://data.terra.bio/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_1e269f04-4347-4188-b060-1dcc69e71d67",
+            "file_ref": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_1e269f04-4347-4188-b060-1dcc69e71d67",
             "file_size": 213021639,
             "is_supplementary": false,
             "reference_assembly": [],
@@ -120,7 +120,7 @@
             "file_id": "8b722e88-8103-49c1-b351-e64fa7c6ab37",
             "file_md5sum": "fNn9e1SovzgOROk3BvH6LQ==",
             "file_name": "307500.merged.matefixed.sorted.markeddups.recal.bam",
-            "file_ref": "drs://data.terra.bio/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
+            "file_ref": "drs://mock_tdr.lan/v1_2ae00e5c-4aef-4a1e-9eca-d8d0747b5348_8b722e88-8103-49c1-b351-e64fa7c6ab37",
             "file_size": 3306845592,
             "is_supplementary": false,
             "reference_assembly": [],

--- a/test/indexer/test_anvil.py
+++ b/test/indexer/test_anvil.py
@@ -5,9 +5,9 @@ from operator import (
     itemgetter,
 )
 from typing import (
+    Type,
     cast,
 )
-import unittest
 from unittest.mock import (
     PropertyMock,
     patch,
@@ -28,6 +28,9 @@ from azul.indexer.document import (
 from azul.logging import (
     configure_test_logging,
 )
+from azul.plugins.repository import (
+    tdr_anvil,
+)
 from azul.plugins.repository.tdr_anvil import (
     BundleEntityType,
     TDRAnvilBundle,
@@ -36,6 +39,9 @@ from azul.plugins.repository.tdr_anvil import (
 from indexer import (
     AnvilCannedBundleTestCase,
     IndexerTestCase,
+)
+from indexer.test_tdr import (
+    TDRPluginTestCase,
 )
 
 
@@ -81,7 +87,11 @@ class AnvilIndexerTestCase(AnvilCannedBundleTestCase, IndexerTestCase):
         }
 
 
-class TestAnvilIndexer(AnvilIndexerTestCase):
+class TestAnvilIndexer(AnvilIndexerTestCase, TDRPluginTestCase[tdr_anvil.Plugin]):
+
+    @classmethod
+    def _plugin_cls(cls) -> Type[tdr_anvil.Plugin]:
+        return tdr_anvil.Plugin
 
     def test_indexing(self):
         self.maxDiff = None
@@ -108,6 +118,17 @@ class TestAnvilIndexer(AnvilIndexerTestCase):
                         self.assertEqual(expected_hits, hits)
                     finally:
                         self.index_service.delete_indices(self.catalog)
+
+    def test_fetch_bundle(self):
+        canned_bundle = self._load_canned_bundle(self.bundle)
+        assert isinstance(canned_bundle, TDRAnvilBundle)
+        self._make_mock_tdr_tables(self.bundle)
+        plugin = self.plugin_for_source_spec(canned_bundle.fqid.source.spec)
+        bundle = plugin.fetch_bundle(self.bundle)
+        assert isinstance(bundle, TDRAnvilBundle)
+        self.assertEqual(canned_bundle.fqid, bundle.fqid)
+        self.assertEqual(canned_bundle.entities, bundle.entities)
+        self.assertEqual(canned_bundle.links, bundle.links)
 
 
 class TestAnvilIndexerWithIndexesSetUp(AnvilIndexerTestCase):
@@ -162,17 +183,3 @@ class TestAnvilIndexerWithIndexesSetUp(AnvilIndexerTestCase):
             DocumentType.contribution: 2,
             **({DocumentType.replica: 2} if config.enable_replicas else {})
         })
-
-    # FIXME: Enable test after the issue with TinyQuery `WITH` has been resolved
-    #        https://github.com/DataBiosphere/azul/issues/5046
-    @unittest.skip('TinyQuery does not support the WITH clause')
-    def test_fetch_bundle(self):
-        canned_bundle = self._load_canned_bundle(self.bundle)
-        assert isinstance(canned_bundle, TDRAnvilBundle)
-        self._make_mock_tdr_tables(self.bundle)
-        plugin = self.plugin_for_source_spec(canned_bundle.fqid.source.spec)
-        bundle = plugin.fetch_bundle(self.bundle)
-        assert isinstance(bundle, TDRAnvilBundle)
-        self.assertEqual(canned_bundle.fqid, bundle.fqid)
-        self.assertEqual(canned_bundle.entities, bundle.entities)
-        self.assertEqual(canned_bundle.links, bundle.links)

--- a/test/indexer/test_tdr.py
+++ b/test/indexer/test_tdr.py
@@ -85,6 +85,7 @@ from azul.terra import (
     TerraCredentialsProvider,
 )
 from azul.types import (
+    AnyJSON,
     JSON,
     JSONs,
     reify,
@@ -232,8 +233,21 @@ class TDRPluginTestCase(TDRTestCase,
         bq.insert_rows(table=table, selected_fields=schema, rows=map(dump_row, rows))
 
     def _bq_schema(self, row: BigQueryRow) -> list[bigquery.SchemaField]:
+
+        def field_type(key: str, value: AnyJSON) -> str:
+            if key == 'version':
+                return 'TIMESTAMP'
+            elif isinstance(value, bool):
+                return 'BOOLEAN'
+            elif isinstance(value, int):
+                return 'INTEGER'
+            else:
+                return 'STRING'
+
         return [
-            bigquery.SchemaField(name=k, field_type='TIMESTAMP' if k == 'version' else 'STRING')
+            bigquery.SchemaField(name=k,
+                                 field_type=field_type(k, v),
+                                 mode='REPEATED' if isinstance(v, list) else 'NULLABLE')
             for k, v in row.items()
         ]
 
@@ -274,41 +288,37 @@ class TestTDRHCAPlugin(DCP2CannedBundleTestCase,
         bundle = self._load_canned_bundle(self.bundle_fqid)
         # Test valid links
         self._test_fetch_bundle(bundle, load_tables=True)
-
-        # FIXME: Rewrite this part to use BigQuery DML statements
-        #        https://github.com/DataBiosphere/azul/issues/5046
-        if False:
-            self._test_invalid_links(bundle)
-
-    def _test_invalid_links(self, bundle: TDRHCABundle):
-        # Directly modify the canned tables to test invalid links not present
-        # in the canned bundle.
-        dataset = self.source.spec.bq_name
-        links_table = self.tinyquery.tables_by_name[dataset + '.links']
-        links_content_column = links_table.columns['content'].values
-        links_content = json.loads(one(links_content_column))
-        link = first(link
-                     for link in links_content['links']
-                     if link['link_type'] == 'supplementary_file_link')
-        # Test invalid entity_type in supplementary_file_link
-        assert link['entity']['entity_type'] == 'project'
-        link['entity']['entity_type'] = 'cell_suspension'
-        # Update table
-        links_content_column[0] = json.dumps(links_content)
-        # Invoke code under test
-        with self.assertRaises(RequirementError):
-            self._test_fetch_bundle(bundle,
-                                    load_tables=False)  # Avoid resetting tables to canned state
-
-        # Undo previous change
-        link['entity']['entity_type'] = 'project'
-        # Test invalid entity_id in supplementary_file_link
-        link['entity']['entity_id'] += '_wrong'
-        # Update table
-        links_content_column[0] = json.dumps(links_content)
-        # Invoke code under test
-        with self.assertRaises(RequirementError):
-            self._test_fetch_bundle(bundle, load_tables=False)
+        # Test invalid links by modifying the canned bundle
+        spec = self.source.spec
+        plugin = self.plugin_for_source_spec(spec)
+        links = one(plugin.tdr.run_sql(f'''
+            SELECT links_id, content
+            FROM {plugin._full_table_name(spec, 'links')}
+        '''))
+        links_id, links_content = links['links_id'], json.loads(links['content'])
+        link = first(
+            link
+            for link in links_content['links']
+            if link['link_type'] == 'supplementary_file_link'
+        )
+        linked_entity = link['entity']
+        assert linked_entity['entity_type'] == 'project', linked_entity
+        bad_link_fields = [
+            {'entity_type': 'cell_suspension'},
+            {'entity_id': linked_entity['entity_id'] + '_wrong'}
+        ]
+        for field in bad_link_fields:
+            link['entity'] = linked_entity | field
+            # Update table with invalid link
+            plugin.tdr.run_sql(f'''
+                UPDATE {plugin._full_table_name(spec, 'links')}
+                SET content = {json.dumps(links_content)!r}
+                WHERE links_id = "{links_id}"
+            ''')
+            # Invoke code under test
+            with self.assertRaises(RequirementError):
+                self._test_fetch_bundle(bundle,
+                                        load_tables=False)  # Avoid resetting tables to canned state
 
     def test_subgraph_stitching(self):
         downstream_uuid = '4426adc5-b3c5-5aab-ab86-51d8ce44dfbe'

